File: app/src/main/java/org/mesonet/app/caching/Cacher.java
package org.mesonet.app.caching;

import android.app.Activity;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.NonNull;

import org.mesonet.app.dependencyinjection.PerActivity;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.realm.Realm;
import io.realm.RealmConfiguration;
import io.realm.RealmModel;
import io.realm.RealmResults;


@PerActivity
public class Cacher
{
    private Realm mRealm;
    private HandlerThread mRealmThread = new HandlerThread("Realm");



    @Inject
    public Cacher(final Activity inContext)
    {
        mRealmThread.start();
        new Handler(mRealmThread.getLooper(), new Handler.Callback() {
            @Override
            public boolean handleMessage(Message message) {

                synchronized (Cacher.this) {
                    Realm.init(inContext);
                    RealmConfiguration configuration = new RealmConfiguration.Builder().schemaVersion(26)
                            .modules(new MesonetRealmModule())
                            .name(inContext.getPackageName())
                            .build();
                    mRealm = Realm.getInstance(configuration);

                    return false;
                }
            }
        }).sendEmptyMessage(0);
    }



    public void SaveToCache(final CacheDataProvider inCacheController)
    {
        new Handler(mRealmThread.getLooper(), new Handler.Callback() {
            @Override
            public boolean handleMessage(Message message) {
                synchronized (Cacher.this) {
                    
                    mRealm.executeTransaction(new Realm.Transaction() {
                        @Override
                        public void execute(@NonNull Realm inRealm) {
                            inRealm.where(inCacheController.GetClass()).findAll().deleteAllFromRealm();
                            inRealm.copyToRealm(inCacheController.ListToCache());
                        }
                    });

                    return false;
                }
            }
        }).sendEmptyMessage(0);
    }



    public <T extends RealmModel, U> void FindAll(final FindAllListener<T, U> inListener)
    {
        final Looper callingLooper = Looper.myLooper();
        new Handler(mRealmThread.getLooper(), new Handler.Callback() {
            @Override
            public boolean handleMessage(Message message) {
                synchronized (Cacher.this) {
                    final U result = inListener.Process(mRealm.where(inListener.GetClass()).findAll());

                    new Handler(callingLooper, new Handler.Callback() {
                        @Override
                        public boolean handleMessage(Message message) {
                            inListener.Found(result);
                            return false;
                        }
                    }).sendEmptyMessage(0);

                    return false;
                }
            }
        }).sendEmptyMessage(0);
    }



    @Override
    public void finalize()
    {
        if(mRealm != null)
            mRealm.close();

        if(mRealmThread != null)
            mRealmThread.quit();
    }



    public interface FindAllListener<T extends RealmModel, U> extends CacheController<T>
    {
        void Found(U inResults);
        U Process(List<T> inResults);
    }



    public interface CacheDataProvider<T extends RealmModel> extends CacheController<T>
    {
        List<T> ListToCache();
    }



    public interface CacheController<T extends RealmModel>
    {
        Class<T> GetClass();
    }
}

File: app/src/main/java/org/mesonet/app/caching/Cacher.kt
package org.mesonet.app.caching

import android.app.Activity
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.os.Message

import org.mesonet.app.dependencyinjection.PerActivity

import java.util.ArrayList

import javax.inject.Inject

import io.realm.Realm
import io.realm.RealmConfiguration
import io.realm.RealmModel
import io.realm.RealmResults
import io.realm.kotlin.where


@PerActivity
class Cacher @Inject
constructor(inContext: Activity) {
    private var mRealm: Realm? = null
    private val mRealmThread = HandlerThread("Realm")


    init {
        mRealmThread.start()
        Handler(mRealmThread.looper, Handler.Callback {
            synchronized(this@Cacher) {
                Realm.init(inContext)
                val configuration = RealmConfiguration.Builder().schemaVersion(26)
                        .modules(MesonetRealmModule())
                        .name(inContext.packageName)
                        .build()
                mRealm = Realm.getInstance(configuration)

                false
            }
        }).sendEmptyMessage(0)
    }


    fun SaveToCache(inCacheController: CacheDataProvider<*>) {
        Handler(mRealmThread.looper, Handler.Callback {
            synchronized(this@Cacher) {

                mRealm?.executeTransaction { inRealm ->
                    inRealm.where<RealmModel>(inCacheController.GetClass() as Class<RealmModel>?).findAll().deleteAllFromRealm()
                    inRealm.copyToRealm(inCacheController.ListToCache())
                }

                false
            }
        }).sendEmptyMessage(0)
    }


    fun <T : RealmModel, U> FindAll(inListener: FindAllListener<T, U>) {
        val callingLooper = Looper.myLooper()
        Handler(mRealmThread.looper, Handler.Callback {
            synchronized(this@Cacher) {
                val result = inListener.Process(mRealm!!.where(inListener.GetClass()).findAll())

                Handler(callingLooper, Handler.Callback {
                    inListener.Found(result)
                    false
                }).sendEmptyMessage(0)

                false
            }
        }).sendEmptyMessage(0)
    }


    fun finalize() {
        if (mRealm != null)
            mRealm!!.close()

        mRealmThread?.quit()
    }


    interface FindAllListener<T : RealmModel, U> : CacheController<T> {
        fun Found(inResults: U)
        fun Process(inResults: MutableList<T>): U
    }


    interface CacheDataProvider<T : RealmModel> : CacheController<T> {
        fun ListToCache(): List<T>
    }


    interface CacheController<T : RealmModel> {
        fun GetClass(): Class<T>
    }
}
